### PR TITLE
fix send_zc api

### PR DESCRIPTION
**Pull request by zhaozg**

### CODE REVIEW

File: nanomsg-ffi.lua [hunk 1]
```diff
@@ -96,6 +96,29 @@ if (nn.VERSION_CURRENT  - nn.VERSION_AGE) == 0 then
 
     // nn_msg_t doesn't exist in nanomsg; it is a metatype anchor for nn.msg
     struct nn_msg_t { void *ptr; size_t size; };
+    
+    const static int NN_SOL_SOCKET = 0 ;
```
Comment (neomantra) [inline nanomsg-ffi.lua]: These "constants" are supposed to be discovered using nn_symbol.  Was that not working?    NN.MSG (defined below) was a special case; are these constants special somehow?

File: nanomsg-ffi.lua [hunk 2]
```diff
@@ -270,6 +292,11 @@ nn.socket = ffi.metatype( 'struct nn_socket_t', {
             if sz < 0 then return nil, libnn.nn_errno() end
             return nn.msg( ptr[0], sz )
         end,
+        
+        getfd = function(s)
```
Comment (neomantra) [inline nanomsg-ffi.lua]: This is a handy helper function, but it should use the discovered constants (`nn.SOL_SOCKET`, etc.).  Perhaps assert that  constants are not defined, either here or above upon library load?

File: nanomsg-ffi.lua [hunk 3]
```diff
@@ -249,9 +271,9 @@ nn.socket = ffi.metatype( 'struct nn_socket_t', {
         -- Otherwise, returns nil, nn_errorno()
         send_zc = function( s, msg, flags )
             flags = flags or 0
-            local sz = libnn.nn_send( s.fd, msg.ptr, nn.MSG, flags )
```
Comment (neomantra) [inline nanomsg-ffi.lua]: I haven't used nanomsg in a while, but this function is intended for messages allocated with `nn_allocmsg`.  Shouldn't those messages have `nn.MSG` passed for them?  The docs for nn_send say that such a message will already be deallocated, so we shouldn't free it ourselves; so rather we clear out the pointer so the `__gc` metamethod doesn't destroy it.

**Comment by neomantra** [comment]
George, thank you for your contribution.  As you can see, I have noted some issues and questions.  Please address them and we'll work on getting your improvements into the library.

File: nanomsg-ffi.lua (hunk 3 above)
Comment (zhaozg) [inline nanomsg-ffi.lua]: As you know,  nbytes = nn_send (s, &msg, NN_MSG, 0); when len=NN_MSG, buf should be address of pointer(& in c), but lua or luajit without this ability,so nanomsg can't deallocated, so I change use msg.size, and free msg manually.

File: nanomsg-ffi.lua (hunk 1 above)
Comment (zhaozg) [inline nanomsg-ffi.lua]: Sure, I missing some understanding of nn_symbol, I will check again and fix.

File: nanomsg-ffi.lua (hunk 2 above)
Comment (zhaozg) [inline nanomsg-ffi.lua]: same above.

File: nanomsg-ffi.lua (hunk 3 above)
Comment (neomantra) [inline nanomsg-ffi.lua]: OK, I think I see the problem now.   You can achieve address-of by sticking the pointer in an array, like so:
```
local voidptr_1_t = ffi.typeof('void*[1]')
local msg_ptr_addr = voidptr_1_t(msg.ptr)
local sz = libnn.nn_send( s.fd, msg_ptr_addr, nn.MSG, flags )
...
```
`voidptr_1_t` should be cached somewhere else in the library.  I put it nearby in the example for clarity.  Can you try it out this way and see how it works?

Comment (zhaozg) [inline nanomsg-ffi.lua]: Hi, it works good, can you fix it in master?

**Pull request closed without merging**